src/pudl/output/ferc1.py

```diff
@@ -1399,3 +1402,341 @@ def remove_intra_table_calculated_values(exploded: pd.DataFrame) -> pd.DataFrame
+        try:
+            # Look up the tags to apply to this node
+            new_tags = tags_df.loc[(self.source_table, self.xbrl_factoid)].to_dict()
+        #  If it has no tags that's fine. Catch the key error and continue.
+        except KeyError:
+            new_tags = {}
+        logger.debug(
+            f"Found {new_tags=} for node {(self.source_table, self.xbrl_factoid)}"
+        )
+
+        # Construct a copy of this node with the new tags applied.
+        new_tags = self.tags | new_tags
+        new_children = [node.add_tags_from_df(tags_df) for node in self.children]
+        return XbrlCalculationTreeFerc1(
+            **self.dict() | {"tags": new_tags, "children": new_children}
+        )
+
+    def propagate_tags(
+        self: Self, tags: dict[str, str] = {}
+    ) -> "XbrlCalculationTreeFerc1":
+        """Propagate metadata tags to all descendants of tagged notes in the tree.
+
+        TODO: Check that there are no inconsistencies between parent & child node tags.
+        """
+        new_tags = self.tags | tags
+        new_children = [node.propagate_tags(new_tags) for node in self.children]
+        return XbrlCalculationTreeFerc1(
+            **self.dict() | {"tags": new_tags, "children": new_children}
+        )
+
+    def to_networkx(self: Self) -> nx.DiGraph:
+        """Convert the tree to a an undirected NetworkX graph.
+
+        Given a node, treat that node as the root of a tree, and construct an undirected
+        :class:`networkx.Graph` representing the node and all of its children. The
+        tuple of strings (source_table, xbrl_factoid) is used as the ID for each node in
+        the graph.
+        """
+        digraph = nx.DiGraph()  # noqa: N806
+        node_id = (self.source_table, self.xbrl_factoid)
+        # Add the current node and all of its attributes:
+        digraph.add_node(
+            node_id,
+            xbrl_factoid_original=self.xbrl_factoid_original,
+            weight=self.weight,
+            tags=self.tags,
+        )
+        for child in self.children:
+            # Add an edge from the current node to each of its children:
+            child_id = (child.source_table, child.xbrl_factoid)
+            digraph.add_edge(node_id, child_id)
+            # Recursively add all the child nodes and edges:
+            child_graph = child.to_networkx()
+            digraph.add_nodes_from(child_graph.nodes)
+            nx.set_node_attributes(digraph, dict(child_graph.nodes))
+            digraph.add_edges_from(child_graph.edges)
+        return digraph
+
+    def to_leafy_meta(self: Self) -> pd.DataFrame:
+        """Convert an XbrlCalculationTree back into a pandas dataframe.
+
+        - Indexed by (source table, xbrl_factoid)
+        - Each family of tags gets its own column
+        - Weights reflect updated / propagated values
+        - FOREST ONLY: Include column indicating the ID of the root node
+        """
+        G = self.to_networkx()  # noqa: N806
```

Comment:
why the single letter variables? is this a networkx convention or something? it's... ya know hard to read.